torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py

```diff
@@ -0,0 +1,111 @@
+        # Project it on the rest and remove it.
+        if i + 1 < num_cols:
+            rest = matrix[:, i + 1 :]
+            rest -= torch.sum(col * rest, dim=0) * col
+
+
+def powerSGD_hook(
+    process_group: object, bucket: dist._GradBucket, rank: int = 1
+) -> torch.futures.Future:
+    """
+    This DDP communication hook implements a simplified PowerSGD gradient compression
+    algorithm described in https://arxiv.org/abs/1905.13727.
+    Once gradient tensors are aggregated across all workers, this hook applies
+    compression as follows:
+    1) Views the input flattened 1D gradient tensor as a square-shaped tensor M with 0 paddings;
+    2) Decomposes M into two low-rank tensors P and Q,
+    such that M = PQ^T, where Q is initialized from a standard normal distribution and orthogonalized;
+    2) Allreduces P;
+    3) Orthogonizes P;
+    4) Compute Q, which is approximately equal to M^TP;
+    5) Allreduces Q;
+    6) Computes M, which is approximately equal to PQ^T.
+    7) Truncates the input tensor to the original length.
+
+    TODO(wayi@): The above procedure does two matmul+allreduce steps per iteration --
+    one left multiplication and one right multiplication.
+    For warm start, can take one such step at a time, and alternate between them.
+
+    Example::
+        PowerSGDState state(process_group, 1)
+        >>> ddp_model.register_comm_hook(state, powerSGD_hook)
+    """
+    group_to_use = process_group if process_group is not None else dist.group.WORLD
+    world_size = (
+        process_group.size() if process_group is not None else dist.get_world_size()
+    )
+
+    # The input tensor is a flattened 1D tensor.
+    input_tensor = bucket.get_tensors()[0]
+    device = input_tensor.device
+    total_length = input_tensor.shape[0]
+
+    # View the input tensor as a 2D square-shape tensor, and pad 0s if necessary.
+    square_side_length = math.ceil(math.sqrt(total_length))
+    padded_total_length = square_side_length ** 2
+    input_tensor.resize_(padded_total_length)
+    input_tensor[total_length:padded_total_length].fill_(0)
+    matrix = input_tensor.view(square_side_length, square_side_length)
+
+    def create_low_rank_tensor(fill_random_values):
+        "Returns a low-rank 2D tensor and the allocated contiguous memory."
+        memory = torch.empty(square_side_length * rank, device=device)
+        if fill_random_values:
+            with torch.random.fork_rng():
+                # The seed makes sure that the initial random values are the same across all the DDP replicas.
+                # Such seed should differ at every step.
+                # Currently use the length of input tensor as the seed, which should be mostly different.
+                # TODO(wayi@): Should read the random seed from the state of this hook provided by the constructor.
+                torch.manual_seed(total_length)
+                memory[:] = torch.randn_like(memory)
```

Comment:
If you are already optimizing for compression speed, this random number generation step seems to quite expensive. For that reason, the `warm start` alternative, which uses state instead of fresh random numbers is faster. You could also consider pre-generate a tensor of random numbers and select chunks from there. 

Note that you should use fresh random numbers here for each call of the hook (Is that what the TODO says?). `total_length` seems to be fixed.

Reply:
> If you are already optimizing for compression speed, this random number generation step seems to quite expensive. For that reason, the `warm start` alternative, which uses state instead of fresh random numbers is faster. You could also consider pre-generate a tensor of random numbers and select chunks from there.

This version is a bit outdated. Actually I added the `torch.device` to `fork_rng()`, so rng only runs on a single GPU core at a time. I did run another experiment to evaluate the speed, and it seems that such random generator is not too slow.

> Note that you should use fresh random numbers here for each call of the hook (Is that what the TODO says?). `total_length` seems to be fixed.

Thanks for the reminder! `total_length` is the length of input tensor, and for now it is different for each call of the hook (at the same iteration), because the models that I am experimenting on has many dense layers of very different sizes. This assumption may not hold true for CNN.

TODO says that I should pass in a random generator, and use a generated random value as the seed every time. This is the first version of PowerSGD for experimentations, and I don't plan to introduce any extra state for the hook for now.

This means that this version of the hook is stateless, and it does not support warm-start or error feedback. I first want to make sure PowerSGD is promising performance-wise in our setup.

Reply:
Make sense. Thanks for the explanation. When interpreting the outcomes of this test, you should keep in mind that using *different* 'projections' of the gradients at different SGD steps is quite crucial. If you use the same random projection every time, there will be differences between the worker's gradients that are never synchronized.